Fix file upload in nextjs-keystone example

marekryb opened the pull request.

marekryb commented:
@flexdinesh There is one unsolved issue. In keystone.ts / generateUrl there is hardcoded port 4000. This will work fine for 'yarn next:dev', but will not work for 'yarn keystone:dev' (port 3000). Normally I would pass this as some env variable, but not sure how to do this in this example without complicating things. Some guidance please.

flexdinesh commented on examples/nextjs-keystone/src/keystone/schema.ts:
```diff
@@ -19,7 +19,7 @@ const permissions = {
   User: list({
     // readonly for demo purpose
-    access: permissions.readOnly,
+    access: allowAll,
```
We probably have to change the permission to `readOnly`. This example is deployable to vercel in one click using the button in the docs by anyone. `allowAll` gives CRUD access to all fields and it's probably not safe for an example deployed on vercel with public write access.

marekryb replied:
Done

flexdinesh commented on examples/nextjs-keystone/keystone.ts:
```diff
@@ -17,5 +17,16 @@ export default withAuth(
+      my_images: {
+        kind: 'local',
+        type: 'image',
+        generateUrl: path => `http://localhost:4000/images${path}`,
```
We might have to set the URL using environment variables. Apps deployed on vercel will have their URL defined in the env variable `process.env.NEXT_PUBLIC_VERCEL_URL`.

flexdinesh commented:
@marekryb re: generateUrl there is hardcoded port 4000

That's a tricky one. I think we could use environment variables to identify which server we're running (keystone or next). Perhaps we could replace the current `next:dev` script with `"next:dev": "PORT=4000  next dev"` and use the `PORT` env variable to conditionally setup the localhost URL.

But this probably adds an unwarranted layer of complexity to a simple example. Maybe we should just create a separate example to showcase file uploads with nextjs-keystone example and leave this example be as simple as it could be.

marekryb commented:
@flexdinesh I added passing public url through env variable (through cross-env as otherwise someone will come and complain that it doesn't work under Win), but vercel url is prioritized.

See how you like the changes. Alternatives are:
1) split this to separate example,
2) have two graphql apis - one plain and one with upload (upload one can be commented out),
3) leave 'processRequest' block as commented alternative and don't change schema at all.

mmachatschek commented:
@borisno2 this fixes the issue with file uploads and graphql-yoga for me.

If someone imports keystone in a different project (workspace setup), you need to make sure to export the `processRequest` from the keystone project and not adding `graphql-upload` to the dependencies of the other project (because of `instanceof` checks in graphql-upload).
